go/libraries/doltcore/sqle/setalgebra/union.go

```diff
@@ -0,0 +1,342 @@
+			// in the case where you have intervals X and Y defined as A < X < B and B <= Y < C the comparison of the
+			// end of X and the start of Y will be -1.  But X includes all the points less than B, Y includes B and the
+			// points up until C.  So the resulting interval Z would be A < Z < C.
+			resIntervToReduce = Interval{in1.nbf, in1.Start, in2.End}
+		} else {
+			// Non overlapping intervals. Create CompositeSet with intervals in sorted order.
+			return CompositeSet{FiniteSet{make(map[hash.Hash]types.Value)}, []Interval{in1, in2}}, nil
+		}
+	} else if intComparison == noOverlapGreater {
+		if in2.End != nil && in1.Start != nil && (in2.End.Inclusive || in1.Start.Inclusive) && in2.End.Val.Equals(in1.Start.Val) {
+			// see above for info no this case
+			resIntervToReduce = Interval{in1.nbf, in2.Start, in1.End}
+		} else {
+			// Non overlapping intervals. Create CompositeSet with intervals in sorted order.
+			return CompositeSet{FiniteSet{make(map[hash.Hash]types.Value)}, []Interval{in2, in1}}, nil
+		}
+	} else if intComparison[start1start2] <= 0 {
+		if intComparison[end1end2] >= 0 {
+			// the first interval wholly contains the second. Return the first.
+			return in1, nil
+		} else {
+			// return an interval with the smallest start point and largest end point
+			resIntervToReduce = Interval{in1.nbf, in1.Start, in2.End}
+		}
+	} else {
+		if intComparison[end1end2] <= 0 {
+			// the second interval wholly contains the first. Return the second.
+			return in2, nil
+		} else {
+			// return an interval with the smallest start point and largest end point
+			resIntervToReduce = Interval{in1.nbf, in2.Start, in1.End}
+		}
+	}
+
+	return simplifyInterval(resIntervToReduce)
+}
+
+// intervalCompositeSetUnion will check the CompositeSet's FiniteSet for points that the new Interval contains
+// and exclude those from the resulting composite and then union the Interval with its existing intervals.
+func intervalCompositeSetUnion(in Interval, cs CompositeSet) (Set, error) {
+	hashToVal := make(map[hash.Hash]types.Value)
+	for h, v := range cs.Set.HashToVal {
+		contained, err := ValueInInterval(in, v)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if !contained {
+			hashToVal[h] = v
+		}
+	}
+
+	intervals, err := unionWithMultipleIntervals(in, cs.Intervals)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hashToVal) == 0 && len(intervals) == 1 {
+		// could possibly be universal set
+		return simplifyInterval(intervals[0])
+	} else {
+		return CompositeSet{FiniteSet{hashToVal}, intervals}, nil
+	}
+}
+
+// unionWithMultipleIntervals takes an interval and a slice of intervals and returns a slice of intervals containing
+// the minimum number of intervals required to represent the union.  The src []Interval argument must be in sorted
+// order and only contain non-overlapping intervals.
+func unionWithMultipleIntervals(in Interval, src []Interval) ([]Interval, error) {
```

Comment:
If this returned a Set (which would be a CompositeSet in practice) then you could use it as a primitive to compose other union operations